File: server/feed3/src/main/java/info/glennengstrand/db/OutboundDAO.java
package info.glennengstrand.db;

import java.util.List;

import org.joda.time.DateTime;

import com.datastax.driver.core.Session;
import com.google.inject.Inject;

import info.glennengstrand.NewsFeedConfiguration;
import info.glennengstrand.api.Outbound;
import info.glennengstrand.util.Link;

public class OutboundDAO extends CassandraDAO<Outbound> {
        
        private static final String OCCURRED_COLUMN = "Occurred";
        private static final String SUBJECT_COLUMN = "Subject";
        private static final String STORY_COLUMN = "Story";

        @Override
        protected String cql(DataOperation op) {
                String retVal = null;
                switch(op) {
                case LOAD:
                        retVal = "select toTimestamp(occurred) as " + OCCURRED_COLUMN + ", " + SUBJECT_COLUMN + ", " + STORY_COLUMN + " from Outbound where participantid = ? order by occurred desc";
                        break;
                case SAVE:
                        retVal = "insert into Outbound (ParticipantID, " + OCCURRED_COLUMN + ", " + SUBJECT_COLUMN + ", " + STORY_COLUMN + ") values (?, now(), ?, ?) using ttl 7776000";
                        break;
                }
                return retVal;
        }
        
        public void create(Outbound value) {
                upsert(b -> {
                        b.setInt(0, Link.extractId(value.getFrom()).intValue());
                        b.setString(1, value.getSubject());
                        b.setString(2,  value.getStory());
                });
        }
        
        public List<Outbound> fetch(long id) {
                return fetchMulti(b -> {
                        b.setInt(0, new Long(id).intValue());
                }, r -> {
                        return new Outbound.OutboundBuilder()
                                        .withFrom(Link.toLink(id))
                                        .withOccurred(DATE_FORMATTER.print(new DateTime(r.getDate(OCCURRED_COLUMN))))
                                        .withSubject(r.getString(SUBJECT_COLUMN))
                                        .withStory(r.getString(STORY_COLUMN))
                                        .build();
                });
        }

        @Inject
        public OutboundDAO(Session session, NewsFeedConfiguration config) {
                super(session, config);
        }
}

File: server/feed6/src/main/scala/info/glennengstrand/news/db/OutboundDAO.scala
package info.glennengstrand.news.db

import info.glennengstrand.news.core.{ ItemDAO, DocumentIdentity }
import info.glennengstrand.news.model.Outbound
import info.glennengstrand.news.Link
import com.datastax.driver.core.{ Session, PreparedStatement, BoundStatement }
import scala.collection.JavaConverters._
import org.elasticsearch.client.RestHighLevelClient
import java.util.UUID
import org.json4s.JsonDSL._
import org.json4s._
import org.json4s.jackson.JsonMethods._

object OutboundItemDAO {
  def apply(db: Session): OutboundItemDAO = {
    val select = "select toTimestamp(occurred) as occurred, subject, story from Outbound where participantid = ? order by occurred desc"
    val upsert = "insert into Outbound (ParticipantID, Occurred, Subject, Story) values (?, now(), ?, ?)"
    new OutboundItemDAO(db.prepare(select), db.prepare(upsert))
  }
}
class OutboundItemDAO(select: PreparedStatement, upsert: PreparedStatement) extends ItemDAO[Outbound] {

  def gets(id: Long)(implicit db: Session): List[Outbound] = {
    val bs = new BoundStatement(select)
    bs.setInt(0, id.toInt)
    val retVal = for {
      r <- db.execute(bs.bind()).iterator().asScala
      val occurred = r.getTimestamp(0)
      val from = id.toInt
      val subject = r.getString(1)
      val story = r.getString(2)
    } yield Outbound(Option(Link.toLink(from.toLong)), Option(occurred), Option(subject), Option(story))
    retVal.toList
  }
  def add(item: Outbound)(implicit db: Session): Outbound = {
    val bs = new BoundStatement(upsert)
    val data = for {
      from <- item.from
      subject <- item.subject
      story <- item.story
    } yield (from, subject, story)
    data.foreach(d => {
      bs.setInt(0, Link.extractId(d._1).toInt)
      bs.setString(1, d._2)
      bs.setString(2, d._3)
    })
    if (data.size > 0) {
      db.execute(bs.bind())
    }
    item
  }

}

class MockOutboundItemDAO extends ItemDAO[Outbound] {

  def gets(id: Long)(implicit db: Session): List[Outbound] = {
    List()
  }
  def add(item: Outbound)(implicit db: Session): Outbound = {
    item
  }
}

class OutboundDocumentDAO(es: RestHighLevelClient) extends ElasticSearchDAO[Outbound] {
  def client: RestHighLevelClient = {
    es
  }
  def identity: DocumentIdentity = {
    DocumentIdentity("feed", "stories", UUID.randomUUID().toString(), "story", "sender")
  }
  override def source(doc: Outbound, key: String): Map[String, Object] = {
    val retVal = for {
      from <- doc.from
      story <- doc.story
    } yield Map("id" -> key, "sender" -> from, "story" -> story)
    if (retVal.size > 0) {
      retVal.head.asInstanceOf[Map[String, Object]]
    } else {
      Map()
    }
  }
}

class MockOutboundDocumentDAO extends ElasticSearchDAO[Outbound] {
  def client: RestHighLevelClient = {
    null
  }
  def identity: DocumentIdentity = {
    DocumentIdentity("feed", "stories", "test", "story", "sender")
  }
  def source(doc: Outbound): Map[String, Object] = {
    Map()
  }
  override def index(doc: Outbound): Unit = {

  }
  override def search(keywords: String): List[Int] = {
    List()
  }
}

